### PR TITLE
Retry PromptNewDeviceName on regex fail, duplicate

### PR DESCRIPTION
**Pull request by patrickxb**

### CODE REVIEW

**Comment by patrickxb** [comment]
@oconnor663 can you please take a look?

File: go/engine/login_provision.go [hunk 1]
```diff
@@ -484,7 +484,33 @@ func (e *LoginProvision) deviceName(ctx *Context) (string, error) {
...
+		devname, err := ctx.ProvisionUI.PromptNewDeviceName(ctx.GetNetContext(), arg)
+		if err != nil {
+			return "", err
+		}
+		if !libkb.CheckDeviceName.F(devname) {
+			arg.ErrorMessage = "Invalid device name. Device names should be " + libkb.CheckDeviceName.Hint
+			continue
+		}
+		duplicate := false
+		for _, name := range names {
```
Comment (maxtaco) [inline go/engine/login_provision.go]: can you make this a set?  It would make me so happy! :)

File: go/engine/login_provision.go [hunk 2]
```diff
@@ -484,7 +484,33 @@ func (e *LoginProvision) deviceName(ctx *Context) (string, error) {
...
+			if devname == name {
+				duplicate = true
+				break
+			}
+		}
+
+		if duplicate {
+			arg.ErrorMessage = fmt.Sprintf("Device name %q already used", devname)
+			continue
```
Comment (oconnor663) [inline go/engine/login_provision.go]: It looks like `go/client/provision_ui.go` duplicates this check. (It also has its own 10x retry loop.) Should we slim down the client-side code for this?

File: go/engine/login_provision.go [hunk 3]
```diff
@@ -484,7 +484,33 @@ func (e *LoginProvision) deviceName(ctx *Context) (string, error) {
...
 	}
-	return ctx.ProvisionUI.PromptNewDeviceName(ctx.GetNetContext(), arg)
+
+	for i := 0; i < 10; i++ {
+		devname, err := ctx.ProvisionUI.PromptNewDeviceName(ctx.GetNetContext(), arg)
+		if err != nil {
+			return "", err
+		}
+		if !libkb.CheckDeviceName.F(devname) {
+			arg.ErrorMessage = "Invalid device name. Device names should be " + libkb.CheckDeviceName.Hint
```
Comment (oconnor663) [inline go/engine/login_provision.go]: Should we add some code in `go/client/provision_ui.go` to print out this error?

File: go/engine/login_provision.go (hunk 1 above)
Comment (patrickxb) [inline go/engine/login_provision.go]: that creates more memory garbage...but ok.

Comment (maxtaco) [inline go/engine/login_provision.go]: fine fine, ok as is

File: go/engine/login_provision.go (hunk 2 above)
Comment (patrickxb) [inline go/engine/login_provision.go]: I assume that the desktop/mobile apps will also retry like provision ui does.  

The point of this PR was just to check what the client sends back and not bail out of the provisioning process if they send back invalid data.  But I didn't want to prompt infinitely, so there's a retry loop.

Comment (oconnor663) [inline go/engine/login_provision.go]: Got it. So as long as the clients are doing the right thing, we should never hit this block?

File: go/engine/login_provision.go (hunk 3 above)
Comment (patrickxb) [inline go/engine/login_provision.go]: ProvisionUI already uses the checker and will print out the error.

File: go/engine/login_provision.go (hunk 2 above)
Comment (patrickxb) [inline go/engine/login_provision.go]: Yes, but (currently) they don't want to worry about the regex, so I believe in the desktop code they do not check the device name against a regex.

**Comment by patrickxb** [comment]
@oconnor663 is this all ok with you?

**Comment by oconnor663** [comment]
:ship:

**Pull request merged**